subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/GroovyJavaJointIncrementalCompilationIntegrationTest.groovy

```diff
@@ -0,0 +1,90 @@
+
+    def setup() {
+        configureGroovyIncrementalCompilation()
+    }
+
+    private static final Map SOURCES = [
+        'G': 'class G { }',
+        'G.changed': 'class G { int i }',
+        'J': 'class J { }',
+        'J.changed': 'class J { int i; }',
+
+        'G_J': 'class G_J extends J { }',
+
+        'J_G': 'class J_G extends G { }',
+
+        'G_G': 'class G_G extends G { }',
+        'J_G_G': 'class J_G_G extends G_G { }',
+    ]
+
+    @Unroll
+    def 'Groovy-Java joint compilation on #scenario'() {
+        given:
+        applyFileSet(initialSet)
+        run "compileGroovy"
+
+        when:
+        applyFileSet(firstChange)
+        run "compileGroovy", "--info"
+
+        then: 'first build'
+        upToDateOrMessage(firstBuildMessage)
+
+        when: 'second build'
+        applyFileSet(secondChange)
+        run "compileGroovy", "--info"
+
+        then:
+        upToDateOrMessage(secondBuildMessage)
+
+        where:
+        scenario                                    | initialSet            | firstChange                   | firstBuildMessage                                        | secondChange                  | secondBuildMessage
+        'Add Java files to Groovy file set'         | ['G']                 | ['G', 'J']                    | 'changes to non-Groovy files'                            | ['G', 'J']                    | 'UP-TO-DATE'
+        'Add Groovy files to Java file set'         | ['J']                 | ['G', 'J']                    | 'unable to get source-classes mapping'                   | ['G', 'J.changed']            | 'changes to non-Groovy files'
+        'Change Java files in joint compilation'    | ['G', 'J']            | ['G', 'J.changed']            | 'changes to non-Groovy files'                            | ['G', 'J.changed']            | 'UP-TO-DATE'
+        'Change Groovy files which Java depends on' | ['G', 'J_G']          | ['G.changed', 'J_G']          | 'unable to find source file of class J_G'                | ['G.changed', 'J_G']          | 'UP-TO-DATE'
+        'Remove Java files in joint compilation'    | ['G', 'J']            | ['G']                         | 'changes to non-Groovy files'                            | ['G', 'G_G']                  | 'Incremental compilation of '
+        'Remove Groovy files in joint compilation'  | ['G', 'J']            | ['J']                         | 'UP-TO-DATE'/*None of the classes needs to be compiled*/ | ['J', 'G']                    | 'unable to get source-classes mapping'
+        'Add Groovy files to joint file set'        | ['G', 'J']            | ['G', 'J', 'G_G']             | 'Incremental compilation of'                             | ['G', 'J', 'G_G']             | 'UP-TO-DATE'
+        'Change root Groovy files '                 | ['G', 'G_G', 'J_G_G'] | ['G.changed', 'G_G', 'J_G_G'] | 'unable to find source file of class J_G_G'              | ['G.changed', 'G_G', 'J_G_G'] | 'UP-TO-DATE'
```

Comment:
Nice solution! If we take this further, I guess we could extract a fixture where one can configure each of those parts. Maybe even be able to create the fixture via a builder pattern.

Anyway, for now this is perfect.